### PR TITLE
kie-issues#946: Action plan - Exploratory phase of the strategy for implementing the E2E test suite for the new DMN Editor - Part 3

### PR DESCRIPTION
**Pull request by ljmotta**
Related to https://github.com/apache/incubator-kie-tools/issues/3906

This PR continues the exploratory phase and continue the work started on https://github.com/apache/incubator-kie-tools/pull/2158

## Description
This PR creates introduce and fix a few things:
 - Use the `transform` over `zoom` due to Playwright compatibility.
 - Set a `y` offset to the viewport.
 - Remove `variable` from Knowledge Source nodes.
 - Add `add[Edge]Waypoint` tests.
 - Add `delete[Edge]Waypoint` tests.
 - Add `invalid[Edge]` tests.
 - Add jsonModel validation for simple cases.

## Improvements for the next interation
- Add Decision Service node tests
- Add property panel tests
- Add resize node tests
- Improve json model fixture and tests

### CODE REVIEW

File: packages/dmn-editor/tests/e2e/__fixtures__/jsonModel/drgElements.ts [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+  Boolean = "boolean",
+  Context = "context",
+  Date = "date",
+  DateTime = "date and time",
+  DateTimeDuration = "days and time duration",
+  Number = "number",
+  String = "string",
+  Time = "time",
+  YearsMonthsDuration = "years and months duration",
+}
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/__fixtures__/jsonModel/drgElements.ts]: I think we have this in `__fixtures__/jsonModel.ts`

File: packages/dmn-editor/tests/e2e/__fixtures__/nodes.ts [hunk 2]
```diff
@@ -173,6 +164,12 @@ export class Nodes {
     return this.get({ name: args.name }).locator("span", { hasText: args.name }).dblclick();
   }
 
+  public async startDraggingEdge(args: { from: string; edgeType: EdgeType }) {
+    await this.select({ name: args.from, position: NodePosition.TOP });
+    await this.get({ name: args.from }).getByTitle(this.getAddEdgeTitle(args.edgeType)).hover();
+    await this.page.mouse.down();
+  }
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/__fixtures__/nodes.ts:172]:
```suggestion
  /**
   * Used for testing invalid edges. No edge is created using this method.
  */
  public async startDraggingEdge(args: { from: string; edgeType: EdgeType }) {
    await this.select({ name: args.from, position: NodePosition.TOP });
    await this.get({ name: args.from }).getByTitle(this.getAddEdgeTitle(args.edgeType)).hover();
    await this.page.mouse.down();
  }
```

File: packages/dmn-editor/tests/e2e/drgRequirements/invalidAuthorityRequirement.spec.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+  test.describe("Invalid edge - Authority Requirement - From Knowledge Source", () => {
+    test.beforeEach(async ({ palette }) => {
+      await palette.dragNewNode({
+        type: NodeType.KNOWLEDGE_SOURCE,
+        targetPosition: { x: 100, y: 100 },
+        thenRenameTo: "Source Node",
+      });
+    });
+
+    test("shouldn't add an Authority Requirement edge from Input Data node to Input Data node", async ({
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/invalidAuthorityRequirement.spec.ts]:
```suggestion
    test("shouldn't add an Authority Requirement edge from Knowledge Source node to Input Data node", async ({
```

File: packages/dmn-editor/tests/e2e/drgRequirements/invalidAuthorityRequirement.spec.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        type: NodeType.INPUT_DATA,
+        targetPosition: { x: 300, y: 100 },
+      });
+
+      await nodes.startDraggingEdge({ from: "Source Node", edgeType: EdgeType.AUTHORITY_REQUIREMENT });
+
+      await expect(nodes.get({ name: DefaultNodeName.INPUT_DATA })).toHaveClass(/.*dimmed/);
+    });
+
+    test("shouldn't add an Authority Requirement edge from Input Data node to Decision Service node", async ({
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/invalidAuthorityRequirement.spec.ts]:
```suggestion
    test("shouldn't add an Authority Requirement edge from Knowledge Source node to Decision Service node", async ({
```

File: packages/dmn-editor/tests/e2e/drgRequirements/invalidAuthorityRequirement.spec.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        type: NodeType.DECISION_SERVICE,
+        targetPosition: { x: 300, y: 100 },
+      });
+
+      await nodes.startDraggingEdge({ from: "Source Node", edgeType: EdgeType.AUTHORITY_REQUIREMENT });
+
+      await expect(nodes.get({ name: DefaultNodeName.DECISION_SERVICE })).toHaveClass(/.*dimmed/);
+    });
+
+    test("shouldn't add an Authority Requirement edge from Input Data node to Group node", async ({
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/invalidAuthorityRequirement.spec.ts]:
```suggestion
    test("shouldn't add an Authority Requirement edge from Knowledge Source node to Group node", async ({
```

File: packages/dmn-editor/tests/e2e/drgRequirements/invalidAuthorityRequirement.spec.ts [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        type: NodeType.GROUP,
+        targetPosition: { x: 300, y: 100 },
+      });
+
+      await nodes.startDraggingEdge({ from: "Source Node", edgeType: EdgeType.AUTHORITY_REQUIREMENT });
+
+      await expect(nodes.get({ name: DefaultNodeName.GROUP })).toHaveClass(/.*dimmed/);
+    });
+
+    test("shouldn't add an Authority Requirement edge from Input Data node to Text Annotation node", async ({
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/invalidAuthorityRequirement.spec.ts]:
```suggestion
    test("shouldn't add an Authority Requirement edge from Knowledge Source node to Text Annotation node", async ({
```

File: packages/dmn-editor/tests/e2e/drgRequirements/invalidKnowledgeRequirement.spec.ts [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+  test.describe("Invalid edge - Knowledge Requirement - From BKM", () => {
+    test.beforeEach(async ({ palette }) => {
+      await palette.dragNewNode({
+        type: NodeType.BKM,
+        targetPosition: { x: 100, y: 100 },
+        thenRenameTo: "Source Node",
+      });
+    });
+
+    test("shouldn't add an Knowledge Requirement edge from Input Data node to Input Data node", async ({
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/invalidKnowledgeRequirement.spec.ts]:
```suggestion
    test("shouldn't add an Knowledge Requirement edge from BKM node to Input Data node", async ({
```

File: packages/dmn-editor/tests/e2e/drgRequirements/invalidKnowledgeRequirement.spec.ts [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        type: NodeType.INPUT_DATA,
+        targetPosition: { x: 300, y: 100 },
+      });
+
+      await nodes.startDraggingEdge({ from: "Source Node", edgeType: EdgeType.KNOWLEDGE_REQUIREMENT });
+
+      await expect(nodes.get({ name: DefaultNodeName.INPUT_DATA })).toHaveClass(/.*dimmed/);
+    });
+
+    test("shouldn't add an Knowledge Requirement edge from Input Data node to Knowledge Source node", async ({
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/invalidKnowledgeRequirement.spec.ts]:
```suggestion
    test("shouldn't add an Knowledge Requirement edge from BKM node to Knowledge Source node", async ({
```

File: packages/dmn-editor/tests/e2e/drgRequirements/invalidKnowledgeRequirement.spec.ts [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        type: NodeType.KNOWLEDGE_SOURCE,
+        targetPosition: { x: 300, y: 100 },
+      });
+
+      await nodes.startDraggingEdge({ from: "Source Node", edgeType: EdgeType.KNOWLEDGE_REQUIREMENT });
+
+      await expect(nodes.get({ name: DefaultNodeName.KNOWLEDGE_SOURCE })).toHaveClass(/.*dimmed/);
+    });
+
+    test("shouldn't add an Knowledge Requirement edge from Input Data node to Group node", async ({
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/invalidKnowledgeRequirement.spec.ts]:
```suggestion
    test("shouldn't add an Knowledge Requirement edge from BKM node to Group node", async ({
```

File: packages/dmn-editor/tests/e2e/drgRequirements/invalidKnowledgeRequirement.spec.ts [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        type: NodeType.GROUP,
+        targetPosition: { x: 300, y: 100 },
+      });
+
+      await nodes.startDraggingEdge({ from: "Source Node", edgeType: EdgeType.KNOWLEDGE_REQUIREMENT });
+
+      await expect(nodes.get({ name: DefaultNodeName.GROUP })).toHaveClass(/.*dimmed/);
+    });
+
+    test("shouldn't add an Knowledge Requirement edge from Input Data node to Text Annotation node", async ({
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/invalidKnowledgeRequirement.spec.ts]:
```suggestion
    test("shouldn't add an Knowledge Requirement edge from BKM node to Text Annotation node", async ({
```

**Review by jomarko — COMMENTED** [review]
I found some parts that we need to address probably. Please take a look when you have a moment and also sorry, probably myself is the author of problematic lines.

**Review by jomarko — APPROVED** [review]
Thank you for updates

**Review by tiagobento — APPROVED** [review]

**Comment by tiagobento** [comment]
@ljmotta Looking at the `jsonModel.ts` fixture you introduced, I have a couple of thoughts:
1. The wait mechanism is flawed, we can do better with a new optional prop called `onModelDebounceStateChanged` only render the JSON once debounce is completed.
2. The `drgElements.ts` file has too much logic. IMHO we shouldn't be abstracting direct access to the JSON model itself. IMHO it's alright to use hard-coded indexes to find stuff in the array, as we expect that the JSON model will be stable.
3. The strategy of using the DOM to make the JSON available to the tests is a clever hack, well done.

**Comment by tiagobento** [comment]
@ljmotta I'm having the exact same problem at apache/incubator-kie-issues#2230, I'll fix it there and this PR should go back to working normally.

**Comment by tiagobento** [comment]
apache/incubator-kie-issues#2230 merged. I already re-triggered the CI here.

**Pull request merged**